chainer/dataset/dataset_mixin.py

```diff
@@ -0,0 +1,46 @@
+class DatasetMixin(object):
+
+    """Default implementation of dataset indexing.
+
+    DatasetMixin provides the :meth:`__getitem__` method using the per-example
+    addressing method :meth:`get_example` given by the implementation. This
+    mixin makes it easy to implement a new dataset that does not support
+    efficient slicing.
+
+    """
+    def __getitem__(self, index):
+        """Returns an example or a sequence of examples.
+
+        It implements the standard Python indexing. It uses the
+        :meth:`get_example` method by default, but it may be overridden by the
+        implementation to improve the slicing performance.
+
+        """
+        if isinstance(index, slice):
+            current, stop, step = index.indices(len(self))
+            ret = []
+            while current < stop and step > 0 or current > stop and step < 0:
+                ret.append(self.get_example(current))
+                current += step
+            return ret
+        else:
+            return self.get_example(index)
```

Comment:
Hmm, I think it is better to check types more thoroughly, but I don't know its impact on the performance. I think if we check whether the index is int, we also have to check if the elements of the slice are int, because they can be arbitrary values. Checking all of them introduces much more overhead.